common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java

```diff
@@ -33,65 +59,120 @@ protected void initializeSharedPreferencesFileManager(String fileName) {
     /**
      * Method responsible for saving tokens contained in the TokenResponse to storage.
      *
-     * @param oAuth2Strategy
+     * @param strategy
      * @param request
      * @param response
      */
     @Override
-    public void saveTokens(OAuth2Strategy oAuth2Strategy, AuthorizationRequest request, TokenResponse response) {
+    public void saveTokens(OAuth2Strategy strategy, AuthorizationRequest request, TokenResponse response) {
 
-        Account account = oAuth2Strategy.createAccount(response);
-        String issuerCacheIdentifier = oAuth2Strategy.getIssuerCacheIdentifier(request);
-        AccessToken accessToken = oAuth2Strategy.getAccessTokenFromResponse(response);
-        RefreshToken refreshToken = oAuth2Strategy.getRefreshTokenFromResponse(response);
+        Account account = strategy.createAccount(response);
+        String issuerCacheIdentifier = strategy.getIssuerCacheIdentifier(request);
+        RefreshToken refreshToken = strategy.getRefreshTokenFromResponse(response);
+
+        ADALTokenCacheItem cacheItem = new ADALTokenCacheItem(strategy, request, response);
 
         //There is more than one valid user identifier for some accounts... AAD Accounts as of this writing have 3
-        ListIterator<String> cacheIds = account.getCacheIdentifiers().listIterator();
+        ListIterator<String> accountCacheIds = account.getCacheIdentifiers().listIterator();
 
-        while (cacheIds.hasNext()) {
+        while (accountCacheIds.hasNext()) {
             //Azure AD Uses Resource and Not Scope... but we didn't override... heads up
-            setItemToCacheForUser(issuerCacheIdentifier, request.getScope(), request.getClientId(), accessToken, refreshToken, cacheIds.next());
+            setItemToCacheForUser(issuerCacheIdentifier, request.getScope(), request.getClientId(), cacheItem, accountCacheIds.next());
+        }
+
+        ListIterator<IShareSingleSignOnState> otherCaches = mSharedSSOCaches.listIterator();
+
+        while (otherCaches.hasNext()) {
+            otherCaches.next().setSingleSignOnState(account, refreshToken);
         }
 
         // TODO: I'd like to know exactly why this is here before I put this back in.... i'm assuming for ADFS v3.
         //setItemToCacheForUser(resource, clientId, result, null);
     }
 
 
-    private void setItemToCacheForUser(final String issuer, final String resource, final String clientId, final AccessToken accessToken, final RefreshToken refreshToken, final String userId) {
+    private void setItemToCacheForUser(final String issuer, final String resource, final String clientId, final ADALTokenCacheItem cacheItem, final String userId) {
+
+        setItem(CacheKey.createCacheKeyForRTEntry(issuer, resource, clientId, userId), cacheItem);
```

Comment:
We should inspect the creation of CacheKeys to see that it supports the objective of #39 -- we should be able to retrieve the tokens for a user even if their upn/email differ